### PR TITLE
Feature/Added actions to profile list items

### PR DESCRIPTION
**Pull request by davidfranquet**
#### 📲 What

Added the following functionality to the Profile screen
- Avatar picker (camera & gallery) this now only saves the picture locally, but need to implement the functionality to save the path on Firebase (https://amidodevelopment.atlassian.net/browse/FARM-497)
- Switch language
- Your farm details
- Invite friends (@harrylusk can you provide us the text to show on the invitation? Now the text is: **Join to the new smart farming app - FarmSmart**
- Create new profile (there's a known bug with the chat flow https://amidodevelopment.atlassian.net/browse/FARM-496)
- Privacy Policy
- Terms of use (missing correct url @harrylusk can you provide us the correct one)

#### 🤔 Why
		
 it's needed

#### ✅ Acceptance criteria

- [ ] Design Review for UI with BA completed. 
- [ ] Manually tested and verified on Android device/emulator (min. Lollipop 5.1)
- [ ] Showcase video / automation test
- [ ] Passing all tests
- [ ] Rebased/merged with latest changes from development and re-tested
- [ ] Removed unsed comments. TODOs must have JIRA for future work.

#### Coding Standards Checklist
- [ ] unit tests 80% coverage on testable code (functions, methods, class)
- [ ] (Architecture) redux state, reducers, actions, middleware defined under lib/redux
- [ ] ui elements defined under lib/ui. 
- [ ] Strings ready for localisation (e.g. defined in lib/utils/strings.dart)
- [ ] Assets (images/icons path) defined in lib/utils/assets.dart
- [ ] Colors defined in lib/utils/colors.dart
- [ ] Dimensions (ie margins padding) defined in lib/utils/dimens.dart
- [ ] TextStyles defined in lib/utils/styles.dart

Recommended Style Guide: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo

#### 🕵️‍♂️ How to test

Notes for QA

#### Reviewers

@farmsmart/amido @farmsmart/wamf

### CODE REVIEW

File: lib/model/bloc/profile/ProfileDetailProvider.dart [hunk 1]
```diff
@@ -156,4 +151,16 @@ class ProfileDetailProvider
...
     _controller.close();
   }
+
+  void _saveProfileImage(File file, ProfileEntity from) async {
+    final directory = await getApplicationDocumentsDirectory();
+
+    final File newImage =
+        await file.copy('${directory.path}/${from.id}_${from.name}${_Constants.avatarPathSuffix}');
+
+    var savedImagePath = newImage.path;
```
Comment (alexanishchenko88) [inline lib/model/bloc/profile/ProfileDetailProvider.dart:161]: Do not create var if you will not use it more than one time. You can do just print(newImage.path);

File: lib/model/bloc/profile/ProfileDetailProvider.dart [hunk 2]
```diff
@@ -156,4 +151,16 @@ class ProfileDetailProvider
...
+
+    //TODO Update image for User
+    print(savedImagePath);
```
Comment (alexanishchenko88) [inline lib/model/bloc/profile/ProfileDetailProvider.dart:164]: Update image for User flow should be implemented in another PR?

File: lib/ui/profile/Profile.dart [hunk 3]
```diff
@@ -1,22 +1,91 @@
...
 import 'SwitchProfileList.dart';
 
 class _LocalisedStrings {
-  static String activeCrops() => Intl.message('Active crops');
-  static String completedCrops() => Intl.message('Completed');
-  static String buttonTitle() => Intl.message('Switch Profile');
+  static activeCrops() => Intl.message('In progress crops');
+
+  static editFarmDetails() => Intl.message('Edit Farm Details');
+
```
Comment (alexanishchenko88) [inline lib/ui/profile/Profile.dart:32]: Why this space within each line?

File: lib/ui/profile/Profile.dart [hunk 4]
```diff
@@ -1,22 +1,91 @@
...
+  static takePictureFromCamera() => Intl.message('Take picture from camera');
+
+  //TODO Add the correct text & Google Play Link
+  static shareText() => Intl.message(
+      'Join to the new smart farming app - FarmSmart  \n https://linktogoogleplay.com');
+}
+
+class _Strings {
+  static final englishAction = "English";
+  static final swahiliAction = "Swahili";
```
Comment (alexanishchenko88) [inline lib/ui/profile/Profile.dart]: What happens if we have swahili as default language, it still showing language names in english right?

File: lib/ui/profile/Profile.dart [hunk 5]
```diff
@@ -348,4 +438,196 @@ class Profile extends StatelessWidget {
       ),
     );
   }
+
+  List<ProfileListItemViewModel> _profileItems(
+    ProfileViewModel viewModel,
+    BuildContext context,
+  ) {
+    List<ProfileListItemViewModel> list = [];
```
Comment (alexanishchenko88) [inline lib/ui/profile/Profile.dart]: Maybe profileItemsList will be better name?

File: lib/ui/profile/Profile.dart [hunk 6]
```diff
@@ -348,4 +438,196 @@ class Profile extends StatelessWidget {
...
+    list.add(ProfileListItemViewModel(
+      title: _LocalisedStrings.createNewProfile(),
+      icon: _Icons.soil,
+      onTap: () => _createNewProfile(viewModel, context),
+      isDestructive: false,
+    ));
+
+    list.add(ProfileListItemViewModel(
+      title: _LocalisedStrings.privacyPolicy(),
+      icon: null,
```
Comment (alexanishchenko88) [inline lib/ui/profile/Profile.dart:478]: What if we define this icon as null by default and then just not inform it? This will let us create view model without specify icon as null

File: lib/ui/profile/Profile.dart [hunk 7]
```diff
@@ -184,15 +261,17 @@ class Profile extends StatelessWidget {
...
+    AsyncSnapshot<ProfileViewModel> snapshot,
+  }) {
     final viewModel = snapshot.data;
     return SafeArea(
       child: SingleChildScrollView(
         child: Column(
           children: <Widget>[
             _buildHeader(context, viewModel),
-            _buildItemList(viewModel),
+            _buildItemList(viewModel, context),
```
Comment (alexanishchenko88) [inline lib/ui/profile/Profile.dart:274]: Maybe we can define this context as private attribute of class and then avoid to pass it on each function?

File: lib/ui/profile/Profile.dart [hunk 8]
```diff
@@ -348,4 +438,196 @@ class Profile extends StatelessWidget {
...
+    );
+  }
+
+  _pickImage(ImagePickerLib.ImageSource imageSource, ProfileViewModel viewModel) {
+    ImagePicker.pickImage(
+      onSuccess: (file) {
+        viewModel.saveProfileImage(file);
+      },
+      onError: (message) {
+        //ignore
```
Comment (alexanishchenko88) [inline lib/ui/profile/Profile.dart:624]: Should be really ignored?

File: lib/ui/profile/Profile.dart [hunk 9]
```diff
@@ -348,4 +438,196 @@ class Profile extends StatelessWidget {
...
+  void _openFarmDetails(ProfileViewModel viewModel, BuildContext context) {
+    NavigationScope.presentModal(
+      context,
+      FarmDetails(
+        viewModel: FarmDetailsViewModel(
+          items: _mapToFarmItemViewModel(viewModel.farmDetails),
+          buttonTitle: _LocalisedStrings.editFarmDetails(),
+          confirm: () {
+            //TODO Should Navigate to Chat here?
+            viewModel.newAccountFlow.run(context);
```
Comment (alexanishchenko88) [inline lib/ui/profile/Profile.dart:519]: Why this viewModel has references to FlowCoordinator newAccountFlow?
I don't think it should be part of viewModel, but of business logic I guess.

File: lib/ui/profile/Profile.dart [hunk 10]
```diff
@@ -348,4 +438,196 @@ class Profile extends StatelessWidget {
...
+      ),
+    );
+  }
+
+  void _createNewProfile(ProfileViewModel viewModel, BuildContext context) {
+    viewModel.newAccountFlow.run(context);
+  }
+
+  void _inviteFriends() async {
+    await Share.share(_LocalisedStrings.shareText());
```
Comment (alexanishchenko88) [inline lib/ui/profile/Profile.dart:531]: Wow, so easy 😄

**Review by alexanishchenko88 — COMMENTED** [review]
Some comments left

File: lib/ui/profile/Profile.dart (hunk 4 above)
Comment (LeeMatthewHiggins) [reply lib/ui/profile/Profile.dart]: The language strings should all be in the native language (so they don't need to run via the translation system) We need to use the translated string for "Swahili"google says its "Kiswahili"

File: lib/ui/profile/Profile.dart (hunk 7 above)
Comment (LeeMatthewHiggins) [reply lib/ui/profile/Profile.dart:274]: nope, don't do that you'll be adding state to a stateless widget

File: lib/ui/profile/Profile.dart (hunk 5 above)
Comment (LeeMatthewHiggins) [reply lib/ui/profile/Profile.dart]: its in the context of _profileItems, I think its ok. maybe items is better, but I think list is fine?

File: lib/ui/profile/Profile.dart (hunk 6 above)
Comment (LeeMatthewHiggins) [reply lib/ui/profile/Profile.dart:478]: 👍

File: lib/ui/profile/Profile.dart (hunk 9 above)
Comment (LeeMatthewHiggins) [reply lib/ui/profile/Profile.dart:519]: it is defined by business logic, and only passed on by the viewmodel. So not too bad. But you are right, this is up for review as its not ideal. But for now it´s the best thing we can do.

File: lib/model/bloc/profile/ProfileDetailProvider.dart (hunk 2 above)
Comment (LeeMatthewHiggins) [reply lib/model/bloc/profile/ProfileDetailProvider.dart:164]: 👍

File: lib/ui/profile/Profile.dart (hunk 3 above)
Comment (LeeMatthewHiggins) [reply lib/ui/profile/Profile.dart:32]: 👍

**Review by LeeMatthewHiggins — CHANGES_REQUESTED** [review]
minor things from Alex

File: lib/ui/profile/Profile.dart (hunk 3 above)
Comment (davidfranquet) [reply lib/ui/profile/Profile.dart:32]: Format code does this

File: lib/model/bloc/profile/ProfileDetailProvider.dart (hunk 2 above)
Comment (davidfranquet) [reply lib/model/bloc/profile/ProfileDetailProvider.dart:164]: yep

File: lib/ui/profile/Profile.dart (hunk 5 above)
Comment (alexanishchenko88) [reply lib/ui/profile/Profile.dart]: With name like items is good to me, just do not call list as list.

File: lib/ui/profile/Profile.dart (hunk 7 above)
Comment (alexanishchenko88) [reply lib/ui/profile/Profile.dart:274]: Ok, then it's good to me

File: lib/model/bloc/profile/ProfileDetailProvider.dart (hunk 1 above)
Comment (davidfranquet) [reply lib/model/bloc/profile/ProfileDetailProvider.dart:161]: this is not finished so I've left this var to let know that the path is the saved one. I'll leave it for now and then the next who takes this (maybe me) can use the newImage.path directly.

Comment (alexanishchenko88) [reply lib/model/bloc/profile/ProfileDetailProvider.dart:161]: Ok, in this case looks good

File: lib/ui/profile/Profile.dart (hunk 6 above)
Comment (davidfranquet) [reply lib/ui/profile/Profile.dart:478]: You're right

File: lib/ui/profile/Profile.dart (hunk 3 above)
Comment (alexanishchenko88) [reply lib/ui/profile/Profile.dart:32]: yeah, I see, because is a function.

File: lib/ui/profile/Profile.dart (hunk 8 above)
Comment (davidfranquet) [reply lib/ui/profile/Profile.dart:624]: Well.. it's not an error it self it's when the user cancel the camera or the gallery so we don't want to do anything

File: lib/ui/profile/Profile.dart (hunk 5 above)
Comment (davidfranquet) [reply lib/ui/profile/Profile.dart]: items is better for me

Comment (davidfranquet) [reply lib/ui/profile/Profile.dart]: done

File: lib/ui/profile/Profile.dart (hunk 6 above)
Comment (davidfranquet) [reply lib/ui/profile/Profile.dart:478]: done

File: lib/ui/profile/Profile.dart (hunk 8 above)
Comment (alexanishchenko88) [reply lib/ui/profile/Profile.dart:624]: But in case it raises not due user interaction, or it will never happen?

Comment (davidfranquet) [reply lib/ui/profile/Profile.dart:624]: it will never happen, the onError callback is called when the user cancel the camera, the gallery or the crop. I will change it to onCancel to avoid confusions

Comment (davidfranquet) [reply lib/ui/profile/Profile.dart:624]: done

**Review by alexanishchenko88 — DISMISSED** [review]

File: lib/ui/profile/Profile.dart (hunk 4 above)
Comment (davidfranquet) [reply lib/ui/profile/Profile.dart]: done

**Review by LeeMatthewHiggins — DISMISSED** [review]
LGTM 👍

**Review by alexanishchenko88 — APPROVED** [review]

**Review by LeeMatthewHiggins — APPROVED** [review]

**Comment by harrylusk** [comment]
Awesome to see that God is using our app! hahaha

**Review by albertmiro — APPROVED** [review]

**Comment by harrylusk** [comment]
@DavidFD005 can text read "Download the FarmSmart app here."
